feat: Pull Request Generator - Support excluding PRs by label (#25755)

cwoebker opened the pull request:
<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Implements the enhancement proposal https://github.com/argoproj/argo-cd/issues/25755: adds an `excludedLabels` filter to the Pull Request generator for the **GitHub, GitLab, Gitea, and Azure DevOps** providers.

Pull requests (merge requests for GitLab) carrying **any** of the listed labels are skipped. `excludedLabels` takes precedence over `labels` — a PR matching both an included and an excluded label is excluded.

Notes:
- **Bitbucket** (Server and Cloud) is intentionally excluded: it has no label
  support at all, so there is nothing to filter on.
- Include/exclude filtering is applied consistently across all four providers via a
  single shared helper.
- Documentation is updated for every affected provider, including Gitea's `labels`
  filter, which was previously implemented but undocumented.
- Unit tests cover the exclude path for each provider.

Closes #25755

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

cwoebker committed:
Add basic test for exludedLabels field

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker committed:
Add excludedLabels field to GitHub pull request generator

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker committed:
Add coverage for list filters in new setup

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker committed:
WIP - fix for unit test

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

nitishfy reviewed (commented):
There are few things missing here and I'd like to get answers of few questions here:

1. why do we need a separate excluded label field when we can achieve the same solution to the problem by not assigning a specific lablel to the PR so the preview environment for that PR doesn't get generated?
2. How do we handle the scenario when a PR contains both included and excluded label? which field is getting priority?

cwoebker committed:
Add review feedback

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker committed:
Update documentation

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

pjiang-dev reviewed (commented):
Nice PR. Do you plan on adding support for other providers as well?
If its not too much additional code we can consider adding it to this one as well

cwoebker committed:
Add review feedback

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker committed:
Fix GitHub test naming

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker committed:
Add implementation for GitLab

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker committed:
Add implementation for AzureDevOps

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker commented:
@pjiang-dev I have added the code for GitLab and AzureDevOps as well in two new commits. Let me know what you think!

ranakan19 reviewed (approved):
Looks Good!
Thanks for making the suggested changes from @nitishfy and @pjiang-dev!

Appreciate further explaining your use case. Initially, I viewed this as a nice-to-have functionality, especially given there is a workaround. But thinking about it further, it is pretty common to have include and exclude feature in filtering-like scenarios, so its a logical extension.

Not a blocker:
I wonder if it would be valuable to rename `labels` to a make it more specific as an `includedLabels` i.e the distinction between IncludeLabels and excludeLabels instead of the current behavior of labels defaulting to includedLabels. (not rooting for the name - includedLabels)

cwoebker committed:
Clarify that exludedLabels takes precedence over labels

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker committed:
Add missing documentation for GitLab and AzureDevOps

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker commented:
> can we support [this](https://github.com/argoproj/argo-cd/issues/26766) too?

You mean adding label exclusion support for Gitea as well? Or would you also like to see the rest of the change requests on https://github.com/argoproj/argo-cd/issues/26766#issuecomment-4049496364? Some of those topics might potentially be better suited for a separate PR as I don't want this one to get too broad.

Happy to keep pushing this and add Gitea support @nitishfy, but can you give me some guidance on whether this will be considered and eventually merged. What are the missing steps needed in your eyes?

cwoebker committed:
Run make codegen

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker committed:
Add implementation for Gitea

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker committed:
Consolidate excluded-labels filtering into a shared helper

Replace the three near-identical per-provider helpers (GitHub, GitLab, Azure DevOps) with a single containsAnyExcludedLabels in utils.go. GitHub now converts labels via getGithubPRLabelNames before filtering, matching the Gitea path. Moves the unit test to utils_test.go.

Signed-off-by: Cecil Wöbker <me@cwoebker.com>

cwoebker commented:
@nitishfy I have added Gitea support and also updated the documentation here. I kept the other issues mentioned in https://github.com/argoproj/argo-cd/issues/26766 out of this PR for now, but happy to propose a fix over there as well - as a separate PR to keep the changes minimal.

I would be thankful if you could take a look on this one and let me know if you have any feedback!

nitishfy commented on applicationset/services/pull_request/github_test.go:
```diff
@@ -90,6 +90,76 @@ func TestGetGitHubPRLabelNames(t *testing.T) {
 	}
 }
 
+func TestGitHubListFiltersLabels(t *testing.T) {
```
```suggestion
func TestGitHubListFiltersLabels(t *testing.T) {
    t.Parallel()
```

nitishfy commented on applicationset/services/pull_request/azure_devops_test.go:
```diff
@@ -72,6 +73,69 @@ func TestListPullRequest(t *testing.T) {
+	t.Parallel()
+	svc, err := NewAzureDevOpsService("", "https://dev.azure.com/", "myorg", "myproject", "myrepo", nil, []string{"stale"})
+	require.NoError(t, err)
+	assert.NotNil(t, svc)
```
svc is always non-nil here. I'd suggest we either assert whether excluded labels actually landed on the struct or not. WDYT?
